src/V3MergeCond.cpp

```diff
@@ -0,0 +1,321 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Merge branches/ternary ?:
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2020 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+// V3BranchMerge's Transformations:
+//
+//    Look for sequences of assignments with ternary conditional on the right
+//    hand side with the same condition:
+//      lhs0 = cond ? then0 : else0;
+//      lhs1 = cond ? then1 : else1;
+//      lhs2 = cond ? then2 : else2;
+//
+//    This seems to be a common pattern and can make the C compiler take a
+//    long time when compiling it with optimization. For us it's easy and fast
+//    to convert this to 'if' statements because we know the pattern is common:
+//      if (cond) {
+//          lhs0 = then0;
+//          lhs1 = then1;
+//          lhs2 = then2;
+//      } else {
+//          lhs0 = else0;
+//          lhs1 = else1;
+//          lhs2 = else2;
+//      }
+//
+//   For 1-bit signals, we consider strength reduced forms to be conditionals,
+//   but only if we already encountered a true conditional we can merge with.
+//   If we did, then act as if:
+//      'lhs = cond & value' is actually 'lhs = cond ? value : 1'd0'
+//      'lhs = cond' is actually 'lhs = cond ? 1'd1 : 1'd0'.
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Global.h"
+#include "V3MergeCond.h"
+#include "V3Stats.h"
+#include "V3Ast.h"
+
+//######################################################################
+
+class CheckMergeableVisitor : public AstNVisitor {
+private:
+    bool m_mergeable;
+
+    // VISITORS
+    virtual void visit(AstNode* nodep) VL_OVERRIDE {
+        if (!m_mergeable || !nodep->isPure()) {
+            m_mergeable = false;
```

Comment:
Please make a method to help debug when something isn't merged, call it wherever you now set false with a reason e.g "impure".

```
void clearMergable(astnode* nodep, const char* reason) {
    if (m_mergable) {
       m_mergable = false;
       UINFO(9, "Clearing mergable on "<<nodep<<" due to "<<reason<<endl);
   }
}
```